sdk/program/src/vote/state/mod.rs

```diff
@@ -479,13 +479,67 @@ impl VoteState {
         }
     }
 
-    /// Deserializes the input `VoteStateVersions` buffer directly into a provided `VoteState` struct
+    /// Deserializes the input `VoteStateVersions` buffer directly into the provided `VoteState`.
     ///
-    /// In a BPF context, V0_23_5 is not supported, but in non-BPF, all versions are supported for
-    /// compatibility with `bincode::deserialize`
+    /// In a SBPF context, V0_23_5 is not supported, but in non-SBPF, all versions are supported for
+    /// compatibility with `bincode::deserialize`.
+    ///
+    /// On success, `vote_state` reflects the state of the input data. On failure, `vote_state` is
+    /// reset to `VoteState::default()`.
     pub fn deserialize_into(
         input: &[u8],
         vote_state: &mut VoteState,
+    ) -> Result<(), InstructionError> {
+        // Rebind vote_state to *mut VoteState so that the &mut binding isn't
+        // accessible anymore, preventing accidental use after this point.
+        let vote_state = vote_state as *mut VoteState;
+
+        // Safety: vote_state is valid to_drop (see drop_in_place() docs). After
+        // dropping, the pointer is treated as uninitialized and only accessed
+        // through ptr::write, which is safe as per drop_in_place docs.
+        unsafe {
+            std::ptr::drop_in_place(vote_state);
+        }
+
+        match VoteState::deserialize_into_ptr(input, vote_state) {
+            Ok(()) => Ok(()),
+            Err(err) => {
+                // Safety:
+                //
+                // Deserialize failed so at this point vote_state is uninitialized. We must write a
+                // new _valid_ value into it or after returning from this function the caller is
+                // left with an uninitialized `&mut VoteState`, which is UB (references must always
+                // be valid).
+                //
+                // This is always safe and doesn't leak memory because deserialize_into_ptr() writes
+                // into the fields that heap alloc only when it returns Ok().
+                unsafe {
+                    vote_state.write(VoteState::default());
+                }
+                Err(err)
+            }
+        }
```

Comment:
maybe you intentionally avoided `.map_err()`?. how about the new guy called `inspect_err()` for conciseness (again imo)?: https://doc.rust-lang.org/std/result/enum.Result.html#method.inspect_err

Reply:
imagine ever using is_err!  And yeah I don't like using map_* or inspect when there are side effects.

In cases like this, I am indeed a match fan, it feels more idiomatic than C style err = ...; if err != 0 { ... }

Reply:
I actually ended up removing the match 😭